waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE TypeApplications #-}
+
+module Wasp.AppSpec.Valid
+  ( validateAppSpec,
+    ValidationError (..),
+    getApp,
+    isAuthEnabled,
+  )
+where
+
+import Data.List (find)
+import Data.Maybe (isJust)
+import Wasp.AppSpec (AppSpec)
+import qualified Wasp.AppSpec as AS
+import Wasp.AppSpec.App (App)
+import qualified Wasp.AppSpec.App as App
+import qualified Wasp.AppSpec.App.Auth as Auth
+import Wasp.AppSpec.Core.Decl (takeDecls)
+import qualified Wasp.AppSpec.Entity as Entity
+import qualified Wasp.AppSpec.Entity.Field as Entity.Field
+import qualified Wasp.AppSpec.Page as Page
+
+data ValidationError = GenericValidationError String
+  deriving (Show, Eq)
+
+validateAppSpec :: AppSpec -> [ValidationError]
+validateAppSpec spec =
+  case validateExactlyOneAppExists spec of
+    Just err -> [err]
+    Nothing ->
+      -- NOTE: We check these only if App exists because they all rely on it existing.
+      concat
+        [ validateAppAuthIsSetIfAnyPageRequiresAuth spec,
+          validateAuthUserEntityHasCorrectFields spec
+        ]
+
+validateExactlyOneAppExists :: AppSpec -> Maybe ValidationError
+validateExactlyOneAppExists spec =
+  case AS.takeDecls @App (AS.decls spec) of
+    [_] -> Nothing
+    apps ->
+      Just $
+        GenericValidationError $
+          "Expected exactly 1 'app' declaration, but found " ++ show (length apps) ++ "."
```

Comment:
Hah true :D. I made it sound nicer, as you suggested!

Reply:
I like the update, the only place it feels slightly awkward is if they do not have any app defined:

```haskell
- GenericValidationError "There can be only one 'app' declaration in the whole app! You have 0."
```

But this is super small and I don't think we need to go too crazy checking between <1 and >1 cases.

Reply:
You mean we could make it sound like just `You must define 'app' declaration`?
Ha true, we could do that to make it sound nicer!
Not sure if worth bothering with it now hmmmm (although I spend similar amount of time writing this message :D).

Reply:
Haha didn't mean to sidetrack. As is still prob fine. To cover all the bases if interested, perhaps: `There should be one 'app' declaration per Wasp app. You have X`?

Reply:
Ok, done @shayneczyzewski ! Since I was doing a change anyway, I made special messages, one for the case when no 'app' is there, another for too many of them, I think it should be very human now.

- You are missing an 'app' declaration in your Wasp app.

- You have more than one 'app' declaration in your Wasp app. You have 2.

Reply:
hah nice, love it! :D Making our compiler dev friendly :D

Reply:
Love it!! :) 👍🏻